neo/SmartContract/Native/Tokens/NeoToken.cs

```diff
@@ -0,0 +1,291 @@
+            DistributeGas(engine, account, state);
+            if (amount.IsZero) return;
+            if (state.Votes.Length == 0) return;
+            foreach (ECPoint pubkey in state.Votes)
+            {
+                StorageItem storage_validator = engine.Service.Snapshot.Storages.GetAndChange(CreateStorageKey(Prefix_Validator, pubkey.ToArray()));
+                ValidatorState state_validator = ValidatorState.FromByteArray(storage_validator.Value);
+                state_validator.Votes += amount;
+                storage_validator.Value = state_validator.ToByteArray();
+            }
+            StorageItem storage_count = engine.Service.Snapshot.Storages.GetAndChange(CreateStorageKey(Prefix_ValidatorsCount));
+            ValidatorsCountState state_count = ValidatorsCountState.FromByteArray(storage_count.Value);
+            state_count.Votes[state.Votes.Length - 1] += amount;
+            storage_count.Value = state_count.ToByteArray();
+        }
+
+        private void DistributeGas(ApplicationEngine engine, UInt160 account, AccountState state)
+        {
+            BigInteger gas = CalculateBonus(engine, state.Balance, state.BalanceHeight, engine.Service.Snapshot.PersistingBlock.Index);
+            state.BalanceHeight = engine.Service.Snapshot.PersistingBlock.Index;
+            GAS.MintTokens(engine, account, gas);
+            engine.Service.Snapshot.Storages.GetAndChange(CreateAccountKey(account)).Value = state.ToByteArray();
+        }
+
+        private BigInteger CalculateBonus(ApplicationEngine engine, BigInteger value, uint start, uint end)
+        {
+            if (value.IsZero || start >= end) return BigInteger.Zero;
+            if (value.Sign < 0) throw new ArgumentOutOfRangeException(nameof(value));
+            uint amount = 0;
+            uint ustart = start / Blockchain.DecrementInterval;
+            if (ustart < Blockchain.GenerationAmount.Length)
+            {
+                uint istart = start % Blockchain.DecrementInterval;
+                uint uend = end / Blockchain.DecrementInterval;
+                uint iend = end % Blockchain.DecrementInterval;
+                if (uend >= Blockchain.GenerationAmount.Length)
+                {
+                    uend = (uint)Blockchain.GenerationAmount.Length;
+                    iend = 0;
+                }
+                if (iend == 0)
+                {
+                    uend--;
+                    iend = Blockchain.DecrementInterval;
+                }
+                while (ustart < uend)
+                {
+                    amount += (Blockchain.DecrementInterval - istart) * Blockchain.GenerationAmount[ustart];
+                    ustart++;
+                    istart = 0;
+                }
+                amount += (iend - istart) * Blockchain.GenerationAmount[ustart];
+            }
+            amount += (uint)(engine.Service.Snapshot.GetSysFeeAmount(end - 1) - (start == 0 ? 0 : engine.Service.Snapshot.GetSysFeeAmount(start - 1)));
```

Comment:
Any chance this cast fails when converting `long` to `uint`?
1 GAS is already 1 x 10^8 = 100,000,000, and uint should be limited to ~4 bilions (2^32).
so, 2^32 / 10^8 ~ 40.... shouldn't 40 GAS or more accumulated over this period break this sum?

Reply:
Yes, we should change this